pkg/builder/devcontainer.go

```diff
@@ -0,0 +1,384 @@
+	}
+
+	b.user = buildOutcome.RemoteUser
+
+	if result.ExitCode != 0 {
+		return errors.New("devcontainer build failed")
+	}
+
+	builderCli, err := b.getBuilderDockerClient()
+	if err != nil {
+		return err
+	}
+
+	//	todo: commit sha from project Repository
+	tag := "latest"
+	imageName := fmt.Sprintf("%s:%s", b.localContainerRegistryServer+"/p-"+b.buildId, tag)
+
+	_, err = builderCli.ContainerCommit(context.Background(), buildOutcome.ContainerId, container.CommitOptions{
+		Reference: imageName,
+	})
+	if err != nil {
+		return err
+	}
+
+	b.buildImageName = imageName
+
+	return nil
+}
+
+func (b *DevcontainerBuilder) readConfiguration() error {
+	projectLogger := b.loggerFactory.CreateProjectLogger(b.project.WorkspaceId, b.project.Name)
+	defer projectLogger.Close()
+
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return err
+	}
+
+	dockerClient := docker.NewDockerClient(docker.DockerClientConfig{
+		ApiClient: cli,
+	})
+
+	cmd := []string{"devcontainer", "read-configuration", "--include-features-configuration", "--include-merged-configuration", "--workspace-folder", "/project"}
+	if b.project.Build.Devcontainer.DevContainerFilePath != "" {
+		cmd = append(cmd, "--config", filepath.Join("/project", b.project.Build.Devcontainer.DevContainerFilePath))
+	}
+
+	execConfig := types.ExecConfig{
+		AttachStdout: true,
+		AttachStderr: true,
+		Cmd:          cmd,
+		Tty:          true,
+	}
+
+	result, err := dockerClient.ExecSync(b.buildId, execConfig, nil)
+	if err != nil {
+		return err
+	}
+
+	if result.ExitCode != 0 {
+		return errors.New("devcontainer read configuration failed")
+	}
+
+	// Convert result.Stdout to string
+	stdoutStr := string(result.StdOut)
+	stdoutStr = strings.TrimSuffix(stdoutStr, "\n")
+
+	// Find the index of the last newline character
+	lastNewline := strings.LastIndex(stdoutStr, "\n")
+
+	// If there is a newline character, slice the string from the index after the newline to the end
+	if lastNewline != -1 {
+		stdoutStr = stdoutStr[lastNewline+1:]
+	}
+
+	// Create a new Root object
+	root := &devcontainer.Root{}
+
+	// Unmarshal the JSON into the Root object
+	err = json.Unmarshal([]byte(stdoutStr), root)
+	if err != nil {
+		return err
+	}
+
+	b.postCreateCommands = append(b.postCreateCommands, root.MergedConfiguration.PostCreateCommands...)
+	b.postStartCommands = append(b.postStartCommands, root.MergedConfiguration.Entrypoints...)
+	b.postStartCommands = append(b.postStartCommands, root.MergedConfiguration.PostStartCommands...)
+
+	return nil
+}
+
+func (b *DevcontainerBuilder) startContainer() error {
+	ctx := context.Background()
+
+	cli, err := client.NewClientWithOpts(client.FromEnv, client.WithAPIVersionNegotiation())
+	if err != nil {
+		return err
+	}
+
+	//	todo builder image from config
+	reader, err := cli.ImagePull(ctx, "daytonaio/workspace-project", image.PullOptions{
+		//	auth for builder image
+		//	RegistryAuth: ,
```

Comment:
The comment seems like a leftover. I suggest we remove it.

Reply:
Builder image should be configurable (and can be on a private registry).
Atm. we are using the universal workspace project image for the builder because it has all the dependencies needed for the builder.
In one of the next iterations we would need to address this (publish the official builder image, and allow the user to configure their own if needed).

Reply:
Agreed, but would still suggest we remove the comment l, or at least at TODO: 
`RegistryAuth: ,` just seems off 😅